### DISCUSSION TITLE
RFC-014: Dev server

### DISCUSSION THREAD

**Discussion by g-chao** [web-infra-dev/rspack#518] [RFC]
### Summary
This RFC contains two parts
1. Rspack cli package, which exposes the command interface to end users
2. Rspack-dev-server package, core functionality of rspack dev server.  
### Motivation
Let users use rspack with a development server that provides live reloading. 
### Dependency
The dev server requires two pre-prerequisites:
1. The core build function from rust side
2. The standardized config from user side.  
### Detailed design
![Detailed Flow](https://user-images.githubusercontent.com/10736839/183785366-0d433978-f6c2-46a3-8be7-18e67084d573.png)
1. Users first execute the dev command with raw rspack config file.
```
program.command('build')
  .description('Rspack build cli')
  .argument('<config-json>', 'rspack config json file path')
  .action(async (options) => {
    // config validation
    // call build function
  });

program.command('dev')
  .description('Rspack build cli')
  .argument('<config-json>', 'rspack config json file path')
  .action(async (options) => {
    // config validation
    // call createServer function
  });
```
2. Config validation
3. Executed core build function
4. Start the dev server based on config and build result. 
```
export function createServer(options): Express {
  
  // get user config

  // create main server  
  const app = express();
    
  // add middlewares, like static files hosting, hmr
  app.use(express.static(directory));

  return app;
}

```
### Webpack
Webpack also has a separated webpack-dev-server and webpack-dev-middleware to support local development; mainly have the following major features. 
1. Both CLI and API interfaces
2. No files are written to disk, rather it handles files in memory
3. If files changed in watch mode, the middleware delays requests until compiling has completed.
4. Supports hot module reload (HMR).
### Existing questions and discussions
1. Express vs Koa, which shall we use to bootstrap the dev server. Or, we can just use native http module. 
2. HMR support is in a separated RFC

**Comment by hardfist** [comment]
I think express is more suitable since webpack-dev-server use express internally, which makes the compatible problem more easy, and express middleware is  more mature and rich than koa.

**Comment by hardfist** [comment]
we may also need to investigate which middlewares should dev-server support internally, like cors、 rewrite、http-proxy and so on， and do we need to separate dev-server to another package like webpack
